### DISCUSSION TITLE
Comparison with FreeRTOS RingBuffer

### DISCUSSION THREAD

**Discussion by lgerconsultant** [rlogiacco/CircularBuffer#44] [Q&A] [answered]
Hi, 
Apologies for an ignorant question, but given FreeRTOS ringbuffer, ( https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos_additions.html#ring-buffers)
 what is the advantage or pros/cons of using this library.

Also I assume this isn't threadsafe ( whereas ringbuffer is), so you'd have to wrap threadsafe idioms around the circularbuffer api?

thx
Larry

**Comment by rlogiacco** [comment, marked as answer]
Sorry, never used either the FreeRTOS ring buffer implementation nor FreeRTOS, so I can't tell you what are the differences. From the introductory statement though I would assume the following should be the differentiators:
* this library doesn't have any additional dependency and does not require FreeRTOS
* this library supports both FIFO and LIFO
* this library memory use depends on the number of items you want to store, not on their dimension
* this library requires contiguous memory to be available for the references only (the items themselves can be in non-contiguous memory)
* this library does not store the items themselves, but only their references, as such you don't have to return the objects to the library
* this library can be used with primitive data types and complex data types
* this library is not thread-safe and doesn't provide any specific feature to _send_ objects, it's not a queue nor a communication channel _per se_, so yes, if you want to use it in a multi-threaded context you have to deal with multi-threading yourself

**Reply by lgerconsultant** [reply]
thx for the feedback
